### ISSUE TITLE
Adding AttentiveFP model 

### ISSUE THREAD

**Issue by thegodone** [akensert/molgraph#4]
One of the SOTA model is AttentiveFP: you can find a good implementation (following the original code) here https://github.com/aimat-lab/gcnn_keras/tree/master/kgcnn/literature/AttentiveFP
here a recent paper showing the power of AttFP models: https://pubs.acs.org/doi/10.1021/acsomega.2c06702

**Comment by akensert** [comment]
@thegodone It is difficult to implement AttentiveFP as separate layers (e.g. `AttentiveFPConv`) as the model has two parts  "Atom embedding" and "Molecule embedding". And furthermore, an "initialization step" is performed before the "atom embedding" part, and an aggregation step is performed between the "atom embedding" and "molecule embedding" part. Put simply, it is not feasible to implement it as a `AttentiveFPConv` like `GCNConv`/`GATConv`/etc. It would be more complex than that. At minimum I would need to have two separate Conv layers like "AtomAttentiveFPConv" and "MolAttentiveFPConv" which is not that nice. And perhaps also another two layers to do the "initialization step" and aggregation step (prior to "AtomAttentiveFPConv" and "MolAttentiveFPConv" respectively).

Check out "Supplementary table 3" of  [this paper](https://pubs.acs.org/doi/10.1021/acs.jmedchem.9b00959) 

However, I would like to implement AttentiveFP as a model! :) Working on it now.

**Comment by akensert** [comment]
Actually, have some ideas. Perhaps `NodeAttentiveFPConv` and `GraphAttentiveFPConv` could be something to implement. The latter could be seen as a more sophisticated way to extract a graph vector representation (compared to e.g. `Readout`). I'm not sure yet what I will do. But at least there will likely be some implementation of AttentiveFP.

**Comment by akensert** [comment]
I've now implemented AttentiveFP, as two separate layer types:

```python

attentive_fp = tf.keras.Sequential([
    tf.keras.Input(type_spec=graph_tensor.unspecific_spec),
    # node message passing ("Atom embedding")
    molgraph.layers.AttentiveFPConv(16, apply_initial_node_projection=True),                   
    molgraph.layers.AttentiveFPConv(16),   
    # ....
    # graph message passing ("Molecule embedding")
    molgraph.layers.AttentiveFPReadout(steps=4),   
])

graph_output = attentive_fp(graph_tensor)
```

The model should closely resemble "Supplementary table 3" of [this paper](https://pubs.acs.org/doi/10.1021/acs.jmedchem.9b00959)

**Comment by thegodone** [comment]
Looks good ideaEnvoyé de mon iPhoneLe 26 mai 2023 à 16:10, Alexander Kensert ***@***.***> a écrit :﻿
I've now implemented AttentiveFP, as two separate layer types:
attentive_fp = tf.keras.Sequential([
    tf.keras.Input(type_spec=graph_tensor.unspecific_spec),
    # node message passing ("Atom embedding")
    molgraph.layers.AttentiveFPConv(16, apply_initial_node_projection=True),                   
    molgraph.layers.AttentiveFPConv(16),   
    # ....
    # graph message passing ("Molecule embedding")
    molgraph.layers.AttentiveFPReadout(steps=4),   
])

graph_output = attentive_fp(graph_tensor)
The model should closely resemble "Supplementary table 3" of this paper

**Comment by thegodone** [comment]
I try to use such examples :
<img width="1297" alt="image" src="https://github.com/akensert/molgraph/assets/1186658/2f58db79-7195-47eb-bdfa-bb0268d96b2e">
but is failed to make similar like those: 
<img width="290" alt="image" src="https://github.com/akensert/molgraph/assets/1186658/a6ca6a63-18a1-4909-b064-5c70625edcfd">

**Comment by thegodone** [comment]
Can you explain the steps in AttentiveFPReadout ? I am able to run code now

**Comment by akensert** [comment]
Good to hear you can run the code. 

(As mentioned before, I think), I needed to split up the `AttentiveFP` into two layers: `AttentiveFPConv` and `AttentiveFPReadout`. The first one does the "Atom embedding" part of the AttentiveFP model, which is pretty much a GATConv directly followed by a GRU update (see paper above, specifically sup. table 3.). The second one does the "Moelcule embedding" part, with this "virtual super node" (see also  sup. table 3). (Feel free to check out the source code of this layer, though the implementations should be fine.)

How about `EdgeConv`, `GINConv`, `MPNNConv`? Can you run those? `EdgeConv` will need an additional `NodeReadout` (see documentation).

I think I might rewrite the `EdgeConv` and `MPNNConv` somewhat. Right now they accept `tie_layer` argument, which I think can be omitted. If the GRU update should be shared between layers, it is better to just have a `gru_cell` argument where the same GRUCell can be passed. This is similar to `AttentiveFPConv`, which accepts a `gru_cell` for shared GRU update.

**Comment by thegodone** [comment]
more specifically, what is the step argument in the read out. In AttFP:
PyG / dgl / kgcnn have those 2 layers AtFPConv and Readout.
normally we have 2 AttFPconv and 2 AttFPReadout in default models so is it
step=2 or do I need to do 2 AttFPReadout layers ?

I did not test yet Edge for GINE example.

I am working to integrate the code in OpenOCHEM for benchmark like we have
with kgcnn.

I see that speed is very good in training.



Le lun. 5 juin 2023 à 09:44, Alexander Kensert ***@***.***> a
écrit :

**Comment by akensert** [comment]
The step argument is the number of aggregation steps: `t` in Sup. table 3. _"Normally, we have 2 AttFPconv and 2 AttFPReadout in default models so default step=2."_, perhaps then `AttentiveFPReadout(steps=2)` would be appropriate.

_"I see that speed is very good in training. which will be very powerfull."_, the speed of my implemented `AttentiveFP`? If so, great!:)

`GTConv` can be replace e.g. `GCNConv` or `GATConv` etc. etc. It is similar to multihead attention (of the Transformer) but instead of attending to all nodes, it attends only to adjacent nodes. Not sure what you ask for regarding the autoencoder, feel free to clarify.

**Comment by thegodone** [comment]
"I see that speed is very good in training. which will be very powerfull.", the speed of my implemented AttentiveFP? If so, great!:) Yes

The step argument is the number of aggregation steps: I think t in Sup. table 3. "Normally, we have 2 AttFPconv and 2 AttFPReadout in default models so default step=2." perhaps then AttentiveFPReadout(steps=2) would be appropriate. Yes I set it like this now.

I am trying the mask it is not working yet ;-)
![image](https://github.com/akensert/molgraph/assets/1186658/53890d8d-0376-465e-976a-b1ad7dd0c286)

I am seen recent transformer autoencoder in graphs maybe could be interesting, but to start, I think the GAE from PyG would be already a good example so GATConv decoder I guess, I don't remember the encoder part. ("GTConv can be replace e.g. GCNConv or GATConv etc. etc. It is similar to multihead attention (of the Transformer) but instead of attending to all nodes, it attends only to adjacent nodes. Not sure what you ask for regarding the autoencoder").

**Comment by akensert** [comment]
"_I am trying the mask it is not working yet ;-)_"

What is the problem? What is the error? Make sure that `inputs` are a `GraphTensor` with nested `tf.Tensor`s (and not nested `tf.RaggedTensor`s). Of course in your masking layer you can take care of it:

```python
def call(self, inputs: GraphTensor):
  inp_orig = inputs
  if isinstance(inputs.node_feature, tf.RaggedTensor):
    inputs = inputs.merge() # Now nested `tf.Tensor`s
  # do your stuff here
  return inp_orig.update({'node_feature: updated_node_feature'}) # back to nested `tf.RaggedTensor`s
```

**Comment by akensert** [comment]
I will try to implement a layer (similar to what you have). And also test it out. 

Basically randomly masking nodes/edges with a frequency of `mask_freq`. And the masking values are zero.
